### PR TITLE
Release branch 3.1.0

### LINKED ISSUE #8701: Disable consider-using-join for non-empty separator
### Current problem

In 2.16.0, consider-using-join was extended to include non-empty separators. This makes pylint complains about the following code:

```py
def test(values: List[str]):
    data = ""
    for value in values:
        data += f"prefix{value}suffix"
    return data
```

However I don't think the suggested refactor is an improvement: prefix and suffix are duplicated and the empty list case need to be handled specifically

```py
def test(values: List[str]):
    if not values:
        return ""
    return "prefix" + "suffixprefix".join(values) + "suffix"
```

The current options are:
 - disable the rule completely. This has the drawback of loosing the previous rule for empty separators.
 - use a "pylint: disable" comment. However having to use this on regular basis creates bad habits and reduce the value of linting

### Desired solution

Would it be possible to have an option to disable the rule only for the non-empty separator case?

### Additional context

_No response_

### LINKED ISSUE #8808: wrong-import-order emits multiple identical messages for multiple imports on single line
### Bug description

The same message is printed 5 times and isn't understandable.
image

### Configuration

_No response_

### Command used

```shell
pylint ./main.py
```

### Pylint output

```shell
main.py:3:0: C0411: standard import "import requests, re, signal, time, timeit, logging" should be placed before "import requests, re, signal, time, timeit, logging" (wrong-import-order)
main.py:3:0: C0411: standard import "import requests, re, signal, time, timeit, logging" should be placed before "import requests, re, signal, time, timeit, logging" (wrong-import-order)
main.py:3:0: C0411: standard import "import requests, re, signal, time, timeit, logging" should be placed before "import requests, re, signal, time, timeit, logging" (wrong-import-order)
main.py:3:0: C0411: standard import "import requests, re, signal, time, timeit, logging" should be placed before "import requests, re, signal, time, timeit, logging" (wrong-import-order)
main.py:3:0: C0411: standard import "import requests, re, signal, time, timeit, logging" should be placed before "import requests, re, signal, time, timeit, logging" (wrong-import-order)
```

### Expected behavior

Only one `main.py:3:0: C0411: standard import "import requests, re, signal, time, timeit, logging" should be placed before "import requests, re, signal, time, timeit, logging" (wrong-import-order)` that makes sense

### Pylint version

```shell
pylint 2.17.4
astroid 2.15.5
Python 3.10.6 (main, May 29 2023, 11:10:38) [GCC 11.3.0]
```

### OS / Environment

Pop!_OS 22.04 LTS

### Additional dependencies

_No response_

### LINKED ISSUE #8827: Extend broad-exception-raised and broad-exception-caught to `except*`
_Originally posted by @clavedeluna in https://github.com/pylint-dev/pylint/issues/7703#issuecomment-1303843158_

***
Expect these examples to raise the same message:

```python
try:
    1 / 0
except Exception:
    pass
```

************* Module a
a.py:3:7: W0718: Catching too general exception Exception (broad-exception-caught)

```python
try:
    1 / 0
except* Exception:
    pass
```

************* Module a
a.py:2:4: W0104: Statement seems to have no effect (pointless-statement)

### LINKED ISSUE #8855: Add `deprecated-attribute` message
### Current problem

Similar to `deprecated-method`, we could have `deprecated-attribute` to warn for uses like this:

```py
>>> sqlite3.version_info
<stdin>:1: DeprecationWarning: version_info is deprecated and will be removed in Python 3.14
```

### Desired solution

`deprecated-attribute` message and then add attributes that were missed from python what's new documents, such as in 3.12:

- sqlite3.version
- sqlite3.version_info
- generator.throw
- agen.throw
- sys.last_type
- sys.last_value
- sys.last_traceback
- calendar.January
- calendar.February

### Additional context

The generator changes might be difficult to make, but the `sys` ones would be easy and high value because they were deprecated only in docs, not with code warnings.

### LINKED ISSUE #8935: Print statistics how many files were checked in verbose.
### Current problem

I am trying to fine tune pylint and when dealing with excludes I see that sometimes too many files were excluded and would like to compare with number of files in a project.

### Desired solution

After pylint completed I would like to have printed something similar to black:
```
black .
2 files reformatted, 13493 files left unchanged.
```

```
pylint *.py **/*.py --verbose
...
13493 files checked.
```

### Additional context

_No response_

### LINKED ISSUE #8947: False negatives for consider-using-min-builtin/consider-using-max-builtin
### Bug description

`consider-using-min-builtin/consider-using-max-builtin` did not catch next situations:

```
value = 5
max_value = 10

if value > max_value:
    max_value = value

if value > float(max_value):
    value = float(max_value)
```

### Configuration

_No response_

### Command used

```shell
pylint --load-plugins=pylint.extensions.code_style --enable=all test.py
```

### Pylint output

```shell
************* Module pylint_test
pylint_test.py:1:0: C0114: Missing module docstring (missing-module-docstring)
pylint_test.py:1:0: C0103: Constant name "value" doesn't conform to UPPER_CASE naming style (invalid-name)
pylint_test.py:2:0: C0103: Constant name "max_value" doesn't conform to UPPER_CASE naming style (invalid-name)
pylint_test.py:5:4: C0103: Constant name "max_value" doesn't conform to UPPER_CASE naming style (invalid-name)

-------------------------------------------------------------------
Your code has been rated at 3.33/10 (previous run: 10.00/10, -6.67)
```

### Expected behavior

First `if` statement should be suggested to be replaced with `max_value = max(max_value, value)`.

Second - with `value = min(value, float(max_value))`.

### Pylint version

```shell
pylint 2.17.4
astroid 2.15.5
Python 3.11.4 (main, Jun 26 2023, 16:57:25) [GCC 7.5.0]
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### LINKED ISSUE #9097: `.pyi` files are not discovered when running on a folder
### Bug description

```python
# asdf/foo.py
foo = 1

# asdf/bar.pyi
foo = 1
```

### Configuration

_No response_

### Command used

```shell
pylint asdf
```

### Pylint output

```
************* Module foo
asdf\foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
asdf\foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
```

### Expected behavior

```
************* Module foo
asdf\foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
asdf\foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
asdf\bar.pyi:1:0: C0104: Disallowed name "foo" (disallowed-name)
asdf\bar.pyi:1:0: C0104: Disallowed name "foo" (disallowed-name)
```

### Pylint version

```shell
pylint 3.0.0a7
astroid 3.0.0
Python 3.11.2 (tags/v3.11.2:878ead1, Feb  7 2023, 16:38:35) [MSC v.1934 64 bit (AMD64)]
```

### OS / Environment

windows 10

### Additional dependencies

_No response_

### LINKED ISSUE #9100: Crash with dataclass enums
### Bug description

When parsing the following ``a.py``:

```python
from dataclasses import dataclass
from enum import Enum


@dataclass
class Something(str, Enum):
	asd: str = 'sdf'
```

### Configuration

_No response_

### Command used

```shell
pylint a.py
```

### Pylint output

```python
<details open>
    <summary>
        pylint crashed with a ``AstroidError`` and with the following stacktrace:
    </summary>


Traceback (most recent call last):
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/lint/pylinter.py", line 788, in _lint_file
    check_astroid_module(module)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/lint/pylinter.py", line 1017, in check_astroid_module
    retval = self._check_astroid_module(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/lint/pylinter.py", line 1069, in _check_astroid_module
    walker.walk(node)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/checkers/base/name_checker/checker.py", line 488, in visit_assignname
    if utils.is_enum_member(node) or utils.is_assign_name_annotated_with(
       ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/checkers/utils.py", line 2284, in is_enum_member
    enum_member_objects = frame.locals.get("__members__")[0].items
                          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^
TypeError: 'NoneType' object is not subscriptable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/lint/pylinter.py", line 752, in _lint_files
    self._lint_file(fileitem, module, check_astroid_module)
  File "/somewhere/test_pylint_enum/.venv/lib64/python3.11/site-packages/pylint/lint/pylinter.py", line 790, in _lint_file
    raise astroid.AstroidError from e
astroid.exceptions.AstroidError




</details>
```

### Expected behavior

No crash. Works on 2.17.7.

### Pylint version

```shell
pylint 3.0.0
astroid 3.0.0
Python 3.11.5 (main, Aug 28 2023, 00:00:00) [GCC 13.2.1 20230728 (Red Hat 13.2.1-1)]
```

### OS / Environment

```shell
NAME="Fedora Linux"
VERSION="38 (Workstation Edition)"
```

### Additional dependencies

_No response_

### LINKED ISSUE #9106: False positive inherit-non-class when inheriting a generic Protocol
### Bug description

In the official Protocol PEP, inheriting directly from a protocol is ok: https://peps.python.org/pep-0544/#explicitly-declaring-implementation

However pylint 3 complains about inheriting from a generic protocol such as below:

```python
class Channel(Protocol[T]):
    """A communication channel between nodes that can be closed"""

    async def get(self) -> T:
        """Raises GeneratorExit when no more values are available"""

    async def put(self, value: T):
        ...

    async def close(self) -> None:
        ...


class DirectChannel(Channel[T]):
    # etc..
```

### Configuration

_No response_

### Command used

```shell
pylint .
```

### Pylint output

```python
************* Module uflow.channel
src/uflow/channel.py:115:0: E0239: Inheriting 'Channel[T]', which is not a class. (inherit-non-class)
src/uflow/channel.py:158:0: E0239: Inheriting 'Channel[T]', which is not a class. (inherit-non-class)
```

### Expected behavior

This should not be an error.

### Pylint version

```shell
pylint 3.0.0
astroid 3.0.0
Python 3.10.10 (main, Aug 24 2023, 21:59:21) [GCC 10.2.1 20210110]
```

### OS / Environment

_No response_

### Additional dependencies

_No response_

### LINKED ISSUE #9228: Resolve pylintrc.toml in pylint 3.x
### Current problem

We are trying to migrate to toml format for pylintrc however if we use
pylintrc it does not detect toml format.

We don't want to have couple of hundrets of lines in pyproject.toml that's why we would like to preserve the external configuration.

### Desired solution

Based on this:
https://pylint.pycqa.org/en/latest/user_guide/usage/run.html

Have an option to resolve pylint.toml or pylintrc.toml in addition to pylintrc and .pylintrc.
One alternative is to have the same toml detection as for argument --rcfile.

Current workaround:
PYLINTRC=pylintrc.toml pylint -v
pylint --rcfile=pylintrc.toml -v

Both options are correctly detecting toml files.

### Additional context

_No response_

### LINKED ISSUE #9229: New check: use `yield from`
### Current problem

Yielding item by item from a generator:

```python
for item in generator:
    yield item
```

### Desired solution

It is cleaner and maybe even a little faster to use `yield from`:

```python
yield from generator
```

This should be an easy check to implement.

### Additional context

Note that `yield from` does [a lot more](https://stackoverflow.com/questions/9708902/in-practice-what-are-the-main-uses-for-the-yield-from-syntax-in-python-3-3/26109157#26109157) than just iterate.

### LINKED ISSUE #9334: Add checks for all set-length-modifying methods in modified-iterating-set
### Current problem

`modified-iterating-set` currently only checks for the `.add` and `.remove` methods, but not several other length-modifying methods.

### Desired solution

Add checks for the `.clear`, `.discard`, and `.pop` methods being called on a set which is being iterated over.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Pierre-Sassoulas**
What's new in Pylint 3.1.0?
---------------------------
Release date: 2024-02-25

New Features
------------

- Skip ``consider-using-join`` check for non-empty separators if an ``suggest-join-with-non-empty-separator`` option is set to ``no``.

  Closes #8701 

- Discover ``.pyi`` files when linting.

  These can be ignored with the ``ignore-patterns`` setting.

  Closes #9097 

- Check ``TypeAlias`` and ``TypeVar`` (PEP 695) nodes for ``invalid-name``.

  Refs #9196 

- Support for resolving external toml files named pylintrc.toml and .pylintrc.toml.

  Closes #9228 

- Check for `.clear`, `.discard`, `.pop` and `remove` methods being called on a set while it is being iterated over.

  Closes #9334 

New Checks
----------

- New message `use-yield-from` added to the refactoring checker. This message is emitted when yielding from a loop can be replaced by `yield from`.

  Closes #9229. 

- Added a ``deprecated-attribute`` message to check deprecated attributes in the stdlib.

  Closes #8855 

False Positives Fixed
---------------------

- Fixed false positive for ``inherit-non-class`` for generic Protocols.

  Closes #9106

- Exempt ``TypedDict`` from ``typing_extensions`` from ``too-many-ancestor`` checks.

  Refs #9167 

False Negatives Fixed
---------------------

- Extend broad-exception-raised and broad-exception-caught to except*.

  Closes #8827 

- Fix a false-negative for unnecessary if blocks using a different than expected ordering of arguments.

  Closes #8947. 

Other Bug Fixes
---------------

- Improve the message provided for wrong-import-order check.  Instead of the import statement ("import x"), the message now specifies the import that is out of order and which imports should come after it.  As reported in the issue, this is particularly helpful if there are multiple imports on a single line that do not follow the PEP8 convention.

  The message will report imports as follows:
  For "import X", it will report "(standard/third party/first party/local) import X"
  For "import X.Y" and "from X import Y", it will report "(standard/third party/first party/local) import X.Y"
  The import category is specified to provide explanation as to why pylint has issued the message and guidence to the developer on how to fix the problem.

  Closes #8808 

Other Changes
-------------

- Print how many files were checked in verbose mode.

  Closes #8935 

- Fix a crash when an enum class which is also decorated with a ``dataclasses.dataclass`` decorator is defined.

  Closes #9100 

Internal Changes
----------------

- Update astroid version to 3.1.0.

  Refs #9457

### CODE REVIEW

File: doc/whatsnew/3/3.1/index.rst [hunk 1]
```diff
@@ -12,5 +12,112 @@
...
+
+  Refs #9196 (`#9196 <https://github.com/pylint-dev/pylint/issues/9196>`_)
+
+- Support for resolving external toml files named pylintrc.toml and .pylintrc.toml.
+
+  Closes #9228 (`#9228 <https://github.com/pylint-dev/pylint/issues/9228>`_)
+
+- New message `use-yield-from` added to the refactoring checker. This message is emitted when yielding from a loop can be replaced by `yield from`.
+
+  Closes #9229. (`#9229 <https://github.com/pylint-dev/pylint/issues/9229>`_)
```
Comment (jacobtylerwalls) [inline doc/whatsnew/3/3.1/index.rst:48]: Let's move this under new checks.

File: doc/whatsnew/3/3.1/index.rst [hunk 2]
```diff
@@ -12,5 +12,112 @@
 Summary -- Release highlights
 =============================
 
+In 3.1.0 we added a new default message ``use-yield-from``,
+a ``deprecated-attribute`` message among other bug fixes.
```
Comment (jacobtylerwalls) [inline doc/whatsnew/3/3.1/index.rst:16]:
```suggestion
Two new checks--``use-yield-from``, ``deprecated-attribute``--
and a smattering of bug fixes.
```

**Review by jacobtylerwalls — DISMISSED** [review]
thanks!

File: doc/whatsnew/3/3.1/index.rst [hunk 3]
```diff
@@ -12,5 +12,111 @@
...
 =============================
 
+Two new checks--``use-yield-from``, ``deprecated-attribute``--
+and a smattering of bug fixes.
 
 .. towncrier release notes start
+
+What's new in Pylint 3.1.0?
+---------------------------
+Release date: 2024-02-24
```
Comment (jacobtylerwalls) [inline doc/whatsnew/3/3.1/index.rst:22]:
```suggestion
Release date: 2024-02-25
```

**Comment by Pierre-Sassoulas** [comment]
Got rusty, had to read the release doc 😄

File: doc/whatsnew/3/3.1/index.rst [hunk 4]
```diff
@@ -12,5 +12,111 @@
 Summary -- Release highlights
```
Comment (jacobtylerwalls) [inline doc/whatsnew/3/3.1/index.rst:12]: I think the date just above this line is okay to update now, too? 🤔

**Comment by jacobtylerwalls** [comment]
We should just yank pandas out of the primer. I know we need to make the primer more stable, but it's just wasting carbon and time until then.

File: doc/whatsnew/3/3.1/index.rst (hunk 4 above)
Comment (Pierre-Sassoulas) [reply doc/whatsnew/3/3.1/index.rst:12]: Hurg, right 🤦

**Review by jacobtylerwalls — APPROVED** [review]
🎉

**Comment by Pierre-Sassoulas** [comment]
Could it be fixed by #7738 ?

**Comment by jacobtylerwalls** [comment]
I don't know if it's due to astroid caching problems or we're not using `--source-roots` because we don't know the layout of the pandas package well, or what, but I do know we're linting a bunch of test directories. I'll start with just limiting the priming of pandas to `pandas.core`?

**Pull request merged**